### PR TITLE
improve: Hide code by default for Markdown cells

### LINKED ISSUE #2389: Hide markdown duplication in 'edit' view.
### Description

I did a search but surprisingly could not see an existing issue for this. When editing a notebook, it would be nice to not see all of the markdown written out twice.

(This is kind of detail that makes Marimo a hard sell to my coworkers.)

I appreciate that you're getting bombarded with a lot of options for when to show/hide each individual widget in each mode. I'm sorry to be adding to that! :D

### Suggested solution

A Jupyter-like solution in which the display switches between the underlying markdown and the rendered HTML, when you click on it to edit it.

I appreciate that Marimo is a bit unusual in that it has *three* representations of Markdown (the underlying Python), which may complicate this further. I think the underlying Python is kind of an implementation detail of the serialization format though, maybe no-one ever needs to worry about that...

### Alternative

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wasimsandhu**
## 📝 Summary

Create Markdown cells on the frontend with `hide_code` set to `true` by default. Closes #2389.

## 🔍 Description of Changes

With @akshayka's change in #2504, this makes it so that the code editor for a Markdown cell is shown when editing, but hidden when the cell is not focused.

## 📋 Checklist

- [x] I have read the [contributor guidelines](https://github.com/marimo-team/marimo/blob/main/CONTRIBUTING.md).
- [x] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://discord.gg/JE7nhX6mD8), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [ ] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

### CODE REVIEW

**Comment by wasimsandhu** [comment]
While testing this, I noticed that I can't click the Delete button for a hidden Markdown cell anymore. The "show code / hide code" overlay keeps getting triggered instead.
Screenshot 2024-10-06 at 12 02 29 PM

**Comment by akshayka** [comment]
I imagined this being config driven, with new markdown cells defaulting to hide_code=True, but this may work. Let me play with it and see how it feels.

**Comment by wasimsandhu** [comment]
Yeah we should probably just do that for behavioral consistency.

**Comment by akshayka** [comment]
> While testing this, I noticed that I can't click the Delete button for a hidden Markdown cell anymore. The "show code / hide code" overlay keeps getting triggered instead. Tried messing around with the z-index but couldn't get it to work.

Oh that must be a consequence of #2504 ...

**Review by mscolnick — COMMENTED** [review]
i love the few line changes like this that bring such good QOL improvements.

File: frontend/src/core/cells/cells.ts [hunk 1]
```diff
@@ -241,6 +242,10 @@ const {
       }
     }
 
+    const isMarkdown = newCellCode
+      ? LanguageAdapters.markdown().isSupported(newCellCode)
```
Comment (mscolnick) [inline frontend/src/core/cells/cells.ts:246]: isSupported returns true for empty string `""`., maybe we could check its not empty as well, or add a function `isMarkdown`

**Comment by wasimsandhu** [comment]
@akshayka Not sure why I thought this change was on the frontend only. Every new Markdown cell that is created now has `config.hide_code` set to True by default, which will get serialized like normal in the notebook file.

File: frontend/src/components/editor/cell/code/cell-editor.tsx [hunk 2]
```diff
@@ -335,17 +336,26 @@ const CellEditorInternal = ({
     };
   }, [editorViewRef]);
 
-  const temporarilyShowCode = async () => {
+  const temporarilyShowCode = useCallback(async () => {
     if (hidden) {
       updateCellConfig({ cellId, config: { hide_code: false } });
       editorViewRef.current?.focus();
-      editorViewParentRef.current?.addEventListener(
+      editorViewRef.current?.contentDOM.addEventListener(
```
Comment (akshayka) [inline frontend/src/components/editor/cell/code/cell-editor.tsx:343]: Hmm, looks like my fix in #2508 got reverted. Is this intentional?

Comment (akshayka) [reply frontend/src/components/editor/cell/code/cell-editor.tsx:343]: Following up with a fix, `blur -> focusout`

Comment (wasimsandhu) [reply frontend/src/components/editor/cell/code/cell-editor.tsx:343]: Yes sorry, I left a comment on that PR but forgot to put one here. And thanks, that did the trick!

File: frontend/src/components/editor/cell/code/cell-editor.tsx [hunk 3]
```diff
@@ -335,17 +336,26 @@ const CellEditorInternal = ({
...
         { once: true },
       );
     }
-  };
+  }, [hidden, cellId, updateCellConfig, editorViewRef]);
+
+  // For a newly created Markdown cell, which defaults to
+  // hidden code, we temporarily show the code editor
+  useEffect(() => {
+    if (!(isMarkdown(code) && code === `mo.md(r"""\n""")`)) {
```
Comment (mscolnick) [inline frontend/src/components/editor/cell/code/cell-editor.tsx:354]: can we do `if (code === new MarkdownLanguageAdapter().defaultCode)`

File: frontend/src/components/editor/cell/code/cell-editor.tsx [hunk 4]
```diff
@@ -335,17 +336,26 @@ const CellEditorInternal = ({
...
+      return;
+    }
+    temporarilyShowCode();
+  }, [code, temporarilyShowCode]);
```
Comment (mscolnick) [inline frontend/src/components/editor/cell/code/cell-editor.tsx:358]: i might be better to just remove `temporarilyShowCode` and add the eslint disable. clearer that it only react to code changes

File: frontend/src/components/editor/cell/code/cell-editor.tsx (hunk 2 above)
Comment (wasimsandhu) [reply frontend/src/components/editor/cell/code/cell-editor.tsx]: Well actually, it looks like you can't delete cells again...

**Review by mscolnick — APPROVED** [review]

**Pull request merged**